hsm/m0hsm.c

```diff
@@ -0,0 +1,748 @@
+		return -1LL;
+
+	return ll;
+}
+
+/**
+ * Read FID in the format [hi:]lo.
+ * Return how many numbers are read.
+ */
+int read_fid(const char *s, struct m0_uint128 *fid)
+{
+	int res;
+	long long hi, lo;
+
+	res = sscanf(s, "%lli:%lli", &hi, &lo);
+	if (res == 1) {
+		fid->u_lo = hi;
+	} else if (res == 2) {
+		fid->u_hi = hi;
+		fid->u_lo = lo;
+	}
+		
+	return res;
+}
+
+static int parse_copy_subopt(char *opts, enum hsm_cp_flags *flags)
+{
+	enum copy_opt {
+		OPT_MOVE = 0,
+		OPT_KEEP_PREV_VERS = 1,
+		OPT_WRITE_TO_DEST = 2,
+	};
+	char *const options[] = {
+		[OPT_MOVE]	     = "mv",
+		[OPT_KEEP_PREV_VERS] = "keep_prev",
+		[OPT_WRITE_TO_DEST]  = "w2dest",
+		NULL,
+	};
+	char *subopts = opts;
+	char *value;
+
+	while (*subopts != '\0') {
+		switch (getsubopt(&subopts, options, &value)) {
+		case OPT_MOVE:
+			*flags |= HSM_MOVE;
+			break;
+
+		case OPT_KEEP_PREV_VERS:
+			*flags |= HSM_KEEP_OLD_VERS;
+			break;
+
+		case OPT_WRITE_TO_DEST:
+			*flags |= HSM_WRITE_TO_DEST;
+			break;
+                default:
+                        fprintf(stderr, "Unexpected option: %s\n", value);
+			return -EINVAL;
+		}
+	}
+	return 0;
+}
+
+static int parse_release_subopt(char *opts, enum hsm_rls_flags *flags)
+{
+	enum release_opt {
+		OPT_KEEP_LATEST = 0,
+	};
+	char *const options[] = {
+		[OPT_KEEP_LATEST] = "keep_latest",
+		NULL,
+	};
+	char *subopts = opts;
+	char *value;
+
+	while (*subopts != '\0') {
+		switch (getsubopt(&subopts, options, &value)) {
+		case OPT_KEEP_LATEST:
+			*flags |= HSM_KEEP_LATEST;
+			break;
+
+                default:
+                        fprintf(stderr, "Unexpected option: %s\n", value);
+			return -EINVAL;
+		}
+	}
+	return 0;
+}
+
+static const struct option option_tab[] = {
+	{"quiet", no_argument, NULL, 'q'},
+	{"verbose", required_argument, NULL, 'v'},
+};
+#define SHORT_OPT "qv"
+
+static int parse_cmd_options(int argc, char **argv)
+{
+	int c, option_index = 0;
+
+	while ((c = getopt_long(argc, argv, SHORT_OPT, option_tab,
+				&option_index)) != -1) {
+		switch (c) {
+		case 'q':
+			if (hsm_options.trace_level > LOG_NONE)
+				hsm_options.trace_level--;
+			break;
+		case 'v':
+			if (hsm_options.trace_level < LOG_DEBUG)
+				hsm_options.trace_level++;
+			break;
+		case ':':
+		case '?':
+		default:
+			return -EINVAL;
+		}
+	}
+	return 0;
+}
+
+/* test functions hidden in m0hsm_api */
+int m0hsm_test_write(struct m0_uint128 id, off_t offset, size_t len, int seed);
+int m0hsm_test_read(struct m0_uint128 id, off_t offset, size_t len);
+
+static int run_cmd(int argc, char **argv)
+{
+	struct m0_uint128 id;
+	const char *action;
+	int rc = 0;
+
+	if (!strcmp(argv[optind], "help")) {
+		usage();
+		return 0;
+	}
+
+	/* expect at least cmd <action> <fid> */
+	if (optind > argc - 2) {
+		usage();
+		return -1;
+	}
+
+	action = argv[optind];
+
+	optind++;
+	id = M0_ID_APP;
+	rc = read_fid(argv[optind], &id);
+	if (rc <= 0) {
+		usage();
+		return -1;
+	}
+	optind++;
+
+	if (!strcmp(action, "create")) {
```

Comment:
int strcmp(const char *s1, const char *s2);

according to our coding style, this should be written as:
```
if (strcmp(action, "create") == 0) {
```

Reply:
Yes, will fix.

Reply:
[nit,optional] Ideally [`m0_streq()`](https://github.com/Seagate/cortx-motr/blob/main/lib/string.h#L34) would be used here.